src/v/cloud_storage/materialized_resources.cc

```diff
@@ -115,84 +105,159 @@ ss::future<> materialized_resources::start() {
-    return static_cast<size_t>(_max_segment_readers_per_shard().value_or(
-      _max_partitions_per_shard() * default_reader_factor));
+    // TODO: implement override logic
+    return memory_groups::tiered_storage_max_memory()
+           / projected_remote_segment_reader_memory_usage();
 }
 
 size_t materialized_resources::max_partition_readers() const {
-    return static_cast<size_t>(_max_partition_readers_per_shard().value_or(
-      _max_partitions_per_shard() * default_reader_factor));
+    // TODO: fixme
+    return memory_groups::tiered_storage_max_memory()
+           / projected_remote_partition_reader_memory_usage();
 }
 
 size_t materialized_resources::max_segments() const {
-    return static_cast<size_t>(_max_segments_per_shard().value_or(
-      _max_partitions_per_shard() * default_segment_factor));
+    // TODO: fixme
+    return memory_groups::tiered_storage_max_memory()
+           / projected_remote_segment_memory_usage();
 }
 
 size_t materialized_resources::current_segment_readers() const {
-    return _segment_reader_units.outstanding();
+    size_t n = 0;
+    for (const auto& s : _materialized) {
+        n += s.readers.size();
+    }
+    return n;
 }
 
 size_t materialized_resources::current_partition_readers() const {
-    return _partition_reader_units.outstanding();
+    // TODO: fixme
+    return 0;
 }
 
 size_t materialized_resources::current_segments() const {
-    return _segment_units.outstanding();
+    return _materialized.size();
 }
 
 void materialized_resources::register_segment(materialized_segment_state& s) {
+    // We can only estimate the actual memory usage after the object is
+    // created. We're using projected memory useage to actually acquire the
+    // units. To fix this we need to check the acquired units and adjust
+    // them. If the units overshoot we need to return some. Otherwise we
+    // need to adopt extra units. When this happens the semaphore counter
+    // may go negative. This means that we used more memory than needed.
+    // In this case we need to mark the segment using 'overcommit' flag
+    // so next eviction will remove it.
+    auto estimate = s.segment->estimate_memory_use();
+    auto units = s._units.count();
+    if (units > estimate) {
+        vlog(
+          cst_log.debug,
+          "Returning extra units. Current: {}, extimate: {}",
+          units,
+          estimate);
+        s._units.return_units(units - estimate);
+    } else {
+        vlog(
+          cst_log.debug,
+          "Adopting extra units. Current: {}, extimate: {}",
+          units,
+          estimate);
+        auto tr = _mem_units.take(estimate - units);
+        s._units.adopt(std::move(tr.units));
+        if (tr.checkpoint_hint) {
+            // We hit the memory limit and the semaphore counter
+            // went below zero. We need to mark this segment as a
+            // candidate for eviction.
+            s.overcommit = true;
```

Comment:
Is this flag used anywhere yet?

Reply:
Initially, I used this flag to prioritize the objects for which the flag is set. But this was removed during code review. Still, I think it makes sense to keep the flag because I'm planning to use it in the followup.